File: packages/go/internal/seguranca/ataques/ataque_canonicalizacao_token.go
package ataques

import (
	"fmt"
	"strings"

	"criptografia/internal/seguranca"
)

// AtaqueCanonicalizacaoToken verifica se um mesmo token aceita múltiplas
// representações textuais válidas (base64url não-canônico / token smuggling).
type AtaqueCanonicalizacaoToken struct{}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueCanonicalizacaoToken) Nome() string {
	return "Canonicalização do token (base64 não-canônico)"
}

type varianteToken struct {
	nome  string
	valor string
}

// Executar roda o ataque.
func (a *AtaqueCanonicalizacaoToken) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	texto := "MENSAGEM_DE_TESTE_DE_CANONICALIZACAO"
	token, err := alvo.Encrypt([]byte(texto))
	if err != nil {
		return seguranca.ResultadoAtaque{}, err
	}
	prefixo := alvo.Prefixo()
	corpo := token[len(prefixo):]
	meio := len(corpo) / 2

	variantes := []varianteToken{}
	for _, p := range []int{0, meio, len(corpo) - 1} {
		variantes = append(variantes,
			varianteToken{nome: fmt.Sprintf("espaço na posição %d", p), valor: prefixo + corpo[:p] + " " + corpo[p:]},
			varianteToken{nome: fmt.Sprintf("newline na posição %d", p), valor: prefixo + corpo[:p] + "\n" + corpo[p:]},
			varianteToken{nome: fmt.Sprintf("tab na posição %d", p), valor: prefixo + corpo[:p] + "\t" + corpo[p:]},
		)
	}
	alfabetoPadrao := strings.ReplaceAll(strings.ReplaceAll(corpo, "-", "+"), "_", "/")
	variantes = append(variantes,
		varianteToken{nome: "alfabeto padrão (+/)", valor: prefixo + alfabetoPadrao},
		varianteToken{nome: "padding \"=\" extra", valor: prefixo + corpo + "="},
		varianteToken{nome: "caractere inválido no meio", valor: prefixo + corpo[:meio] + "!" + corpo[meio:]},
	)

	aceitas := []string{}
	for _, v := range variantes {
		if v.valor == token {
			continue
		}
		if decifrado, err := alvo.Decrypt(v.valor); err == nil && decifrado == texto {
			aceitas = append(aceitas, v.nome)
		}
	}

	if len(aceitas) > 0 {
		return seguranca.ResultadoAtaque{
			NomeAtaque: a.Nome(),
			Vulneravel: true,
			Severidade: seguranca.SevMedia,
			Detalhes: fmt.Sprintf(
				"%d variante(s) textualmente diferente(s) decifram para o mesmo texto: %s",
				len(aceitas), strings.Join(aceitas, "; "),
			),
			Dados: map[string]any{"variantes_aceitas": aceitas},
		}, nil
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: false,
		Severidade: seguranca.SevInfo,
		Detalhes:   fmt.Sprintf("Todas as %d variantes não-canônicas foram rejeitadas", len(variantes)),
	}, nil
}

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueCanonicalizacaoToken.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.CriptografiaAlvo;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Um mesmo token não deveria ter múltiplas representações textuais válidas.
 * Se o decoder de base64url ignora caracteres fora do alfabeto (comportamento
 * padrão do base64_decode não-estrito) ou aceita o alfabeto padrão (+/) no
 * lugar do URL-safe (-_), então strings diferentes decodificam para o MESMO
 * token. Isso é "token smuggling": sistemas que comparam/usam a string do
 * token de formas diferentes (cache, WAF, deduplicação/replay) discordam
 * sobre o que ele significa.
 */
public class AtaqueCanonicalizacaoToken implements AtaqueInterface {
    @Override
    public String nome() {
        return "Canonicalização do token (base64 não-canônico)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        if (!(alvo instanceof CriptografiaAlvo)) {
            throw new SkipAtaqueException("Precisa de base64url_encode/decode do alvo.");
        }

        String texto = "MENSAGEM_DE_TESTE_DE_CANONICALIZACAO";
        String token = alvo.encrypt(texto);
        String prefixo = alvo.prefixo();
        String corpo = token.substring(prefixo.length());
        int meio = corpo.length() / 2;

        Map<String, String> variantes = new LinkedHashMap<>();
        int[] posicoes = {0, meio, corpo.length() - 1};
        for (int p : posicoes) {
            variantes.put("espaço na posição " + p, prefixo + corpo.substring(0, p) + " " + corpo.substring(p));
            variantes.put("newline na posição " + p, prefixo + corpo.substring(0, p) + "\n" + corpo.substring(p));
            variantes.put("tab na posição " + p, prefixo + corpo.substring(0, p) + "\t" + corpo.substring(p));
        }
        variantes.put("alfabeto padrão (+/)", prefixo + corpo.replace('-', '+').replace('_', '/'));
        variantes.put("padding \"=\" extra", prefixo + corpo + "=");
        variantes.put(
                "caractere inválido no meio", prefixo + corpo.substring(0, meio) + "!" + corpo.substring(meio));

        List<String> aceitas = new ArrayList<>();
        for (Map.Entry<String, String> e : variantes.entrySet()) {
            String v = e.getValue();
            if (v.equals(token)) {
                continue;
            }
            try {
                if (alvo.decrypt(v).equals(texto)) {
                    aceitas.add(e.getKey());
                }
            } catch (Exception ex) {
                // esperado
            }
        }

        if (!aceitas.isEmpty()) {
            return new ResultadoAtaque(
                    nome(),
                    true,
                    Severidade.MEDIA,
                    aceitas.size()
                            + " variante(s) textualmente diferente(s) decifram para o mesmo texto: "
                            + String.join("; ", aceitas),
                    Map.of("variantes_aceitas", aceitas));
        }

        return new ResultadoAtaque(
                nome(),
                false,
                Severidade.INFO,
                "Todas as " + variantes.size() + " variantes não-canônicas foram rejeitadas");
    }
}
